src/engine/profiler.cc

```diff
@@ -161,44 +159,47 @@ void Profiler::DumpProfile() {
 
   uint32_t dev_num = cpu_num_ + gpu_num_ + 1;
 
-  for (uint32_t i = 0; i < dev_num; ++i) {
-    const DevStat &d = profile_stat[i];
-    this->EmitPid(&file, d.dev_name, i);
-    file << ",\n";
-  }
-
-  bool first_flag = true;
-  for (uint32_t i = 0; i < dev_num; ++i) {
-    DevStat &d = profile_stat[i];
-    std::lock_guard<std::mutex> lock(d.m_);
-    uint32_t opr_num = d.opr_exec_stats.size();
-
-    for (uint32_t j = 0; j < opr_num; ++j) {
-      const OprExecStat* opr_stat = d.opr_exec_stats[j];
-
-      uint32_t pid = i;
-      uint32_t tid = opr_stat->thread_id;
+  std::shared_ptr<DevStat> pstat = profile_stat_;  // hold a ref count
```

Comment:
why

Reply:
Currently destruction order is not controlled, so we hold a ref count in case the Profgile object is destroyed while this is executing.   It is more relevant for AddOprStat() which may get called from ThreadedEngine::ExecuteOprBlock().  This way, if the pointer isn't null at this time, it doesn't mysteriously become null midway through the function execution.
I am cuyrrently working on the overall singleton problem on another branch.

Reply:
I am making sure that I do not introduce another shutdown crash situation